test-app/tests/integration/components/headless-form-test.gts

```diff
@@ -494,6 +514,837 @@ module('Integration Component headless-form', function (hooks) {
     });
   });
 
+  module('validation', function () {
+    interface FormData {
+      firstName?: string;
+      lastName?: string;
+    }
+
+    const validateFormCallbackSync = ({ firstName }: FormData) => {
+      const firstNameErrors = [];
+      if (firstName == undefined) {
+        firstNameErrors.push({
+          type: 'required',
+          value: firstName,
+          message: 'firstName is required!',
+        });
+      } else {
+        if (firstName.charAt(0).toUpperCase() !== firstName.charAt(0)) {
+          firstNameErrors.push({
+            type: 'uppercase',
+            value: firstName,
+            message: 'firstName must be upper case!',
+          });
+        }
+
+        if (firstName.toLowerCase() === 'foo') {
+          firstNameErrors.push({
+            type: 'notFoo',
+            value: firstName,
+            message: 'Foo is an invalid firstName!',
+          });
+        }
+      }
+
+      return firstNameErrors.length > 0
+        ? { firstName: firstNameErrors }
+        : undefined;
+    };
+
+    const validateFormCallbackAsync = async (data: FormData) => {
+      return await validateFormCallbackSync(data);
+    };
+
+    const validateFieldCallbackSync = (
+      value: string | undefined,
+      field: string
+    ) => {
+      const errors = [];
+      if (value == undefined) {
+        errors.push({
+          type: 'required',
+          value,
+          message: `${field} is required!`,
+        });
+      } else {
+        if (value.charAt(0).toUpperCase() !== value.charAt(0)) {
+          errors.push({
+            type: 'uppercase',
+            value,
+            message: `${field} must be upper case!`,
```

Comment:
Question, total nit: if the first letter starts with a capital, isn't that something like sentence case and not uppercase?

Reply:
Yeah, probably. As a non-native speaker, I was not aware that "sentence case" was a phrase for that. 😅

Given this is only for tests, I would tend to not change this, as it would require changes in a lot of places, unless you insist? :wink: